src/metabase/query_processor/middleware/metrics.clj

```diff
@@ -62,6 +62,18 @@
   [query [_ {:lib/keys [expression-name]} :as expression]]
   (lib/expression query 0 expression-name expression))
 
+(defn- update-metric-query-expression-names
+  [metric-query unique-name-fn]
+  (reduce
+    (fn [metric-query [_ {:lib/keys [expression-name]} :as expression]]
+      (lib/replace-clause
+        metric-query
+        expression
+        (lib/with-expression-name expression (unique-name-fn expression-name))))
+    metric-query
+    ;; Rename in reverse order so dependent expressions are updated in order
+    (reverse (lib/expressions metric-query))))
```

Comment:
In the test case where "qux" depends on "foobar", if "foobar" is renamed first, then `replace-clause` won't find the original "qux" expression since it will have changed to use "foobar_2".

Thinking about this though, I think I need to make a mapping of original-name to new-name and reduce over that.